website/docs/language/state/encryption.html.md

```diff
@@ -0,0 +1,137 @@
+
+# Client-Side Remote State Encryption
+
+State regularly contains mission-critical sensitive information such as credentials.
+
+Several remote state backends already support server-side encryption, but
+
+- many do not
+- some implementations effectively turn encryption into nothing but a proxy for key access control  
+- you need to trust the remote backend operator
+- you need to trust your communication channel to the remote backend
+
+This **experimental feature** lets you encrypt the complete state on the client before transferring it.
+
+Depending on your choice of state crypto provider, no third party ever sees the key.
+
+## Limitations
+
+Client-side state encryption will not work with [enhanced backends](/docs/language/settings/backends/index.html), as 
+those need access to the information in the state to function correctly.
+
+**Client-side remote state encryption is an experimental feature!** Implementation and configuration details are subject to change, 
+and not all remote state backends have been fully tested!
+
+## State Crypto Providers
+
+State crypto providers are a transparent encryption layer around the communication with remote state backends.
+
+**This is an experimental feature!**
+
+You can configure state crypto providers by using two environment variables:
+
+- `TF_REMOTE_STATE_ENCRYPTION`: configuration used for encryption and decryption
+- `TF_REMOTE_STATE_DECRYPTION_FALLBACK`: fallback configuration for decryption, tried if decryption fails with the first choice  
+
+You set each of these to a json document with two fields
+
+  * `implementation`: select a _state crypto provider_ by name
+  * `parameters`: configure the _state crypto provider_
+
+to enable client-side remote state encryption. To disable, either do not set the variable at all or set it to a blank value.
+
+Right now, there is only one value for `implementation`: `client-side/AES256-cfb/SHA256`. 
+
+In the future, more state encryption providers may be added, such as:
+
+- asymmetric encryption with RSA public key cryptography
+- key retrieval from [Vault](https://www.vaultproject.io/)
+- ...
+
+### client-side/AES256-cfb/SHA256
+
+This state crypto provider offers pure client-side symmetric encryption. 
+
+The key is not transferred to any third party. Note that this places the burden of key management on you
+and you alone.
+
+Encryption is performed with AES256-cfb, using a fresh random initialization vector every time. Payload integrity
+is verified using a SHA256 hash over the plaintext, which is encrypted with the plaintext.
+
+_Implementation Name:_ `client-side/AES256-cfb/SHA256`
+
+_Parameters:_
+
+- `key`: the 32 byte AES256 key represented in hexadecimal, must be exactly 64 characters, `0-9a-f` only.
+
+Example:
+
+```shell
+export TF_REMOTE_STATE_ENCRYPTION='{"implementation":"client-side/AES256-cfb/SHA256","parameters":{"key":"a0a1a2a3a4a5a6a7a8a9b0b1b2b3b4b5b6b7b8b9c0c1c2c3c4c5c6c7c8c9d0d1"}}'
+```
+
+## State Encryption Lifecycle
```

Comment:
It would be nice, but I think, obviously, beyond the scope of this Pull Request, to create a CLI tool to help with these operations. Something like:

```bash
terraform state encrypt --implementation 'client-side/AES256-cfb/SHA256' --key 'a0a1a2a3a4a5a6a7a8a9b0b1b2b3b4b5b6b7b8b9c0c1c2c3c4c5c6c7c8c9d0d1'
> Encrypting state, implementation 'client-side/AES256-cfb/SHA256', key: a0a1a2a3a4a5a6a7a8a9b0b1b2b3b4b5b6b7b8b9c0c1c2c3c4c5c6c7c8c9d0d1
> Done

terraform state encrypt
> Encryption Implementations:
> 1. (Recommended) Client-side, AES256, CFB mode, SHA256 integrity check
> 2. Client-side, DES, ECB mode, no integrity check
> 3. Client-side, Pig Latin
>Which encryption implementation would you like to use (default: 1)?: 1
> Key Generation:
> 1. (Recommended) Generate a new completely random key
> 2. Generate a new key from a password PBKDF2
> 3. Use 4 (From random dice roll) https://xkcd.com/221/
> Which key generation algorithm would you like to use (default: 1)? 3
>
> Encrypting state, implementation 'client-side/AES256-cfb/SHA256', key: 0000000000000000000000000000000000000000000000000000000000000004
> Done
```

Reply:
If my PR gets accepted, I'm likely to build a simple encrypt / decrypt cli for downloaded state files/blobs. It is more or less required so state surgery is still possible.

Reply:
@StephanHCB  I honestly don't think yet another tool would be the best approach, but rather give `terraform state pull` (https://www.terraform.io/docs/cli/commands/state/pull.html) the ability to dump the encrypted remote state locally and unencrypted - in short extend the existing terraform state handling commands to deal with encrypted remote state.